### ISSUE TITLE
Developing a streaming native model for Muaalem

### ISSUE THREAD

**Issue by obadx** [obadx/quran-muaalem#10]
### **Problem Statement**
We need to develop a new speech model that natively supports streaming (real-time processing) without relying on workarounds to convert offline models. The model should have an internal memory mechanism similar to sequence models (e.g., LSTMs, GRUs, RNNs) and should be pretrained in a self-supervised manner on a large Arabic speech corpus.

### **Objectives**
- **Architecture:** Streaming-native, with internal memory (e.g., using causal conformer or recurrent conformer blocks).
- **Model Size:** Under 150M paramerts
- Achieves RTF on the CPU of stmart phones with 3GB of RAM.
- **Pretraining:** Self-supervised pretraining on Arabic speech.
- **Base Architecture:** Preferably based on Conformer, balancing accuracy and efficiency.

---

### Proposed Architecture

<img width="957" height="1024" alt="Image" src="https://github.com/user-attachments/assets/ef66c3b9-6d33-4ada-9c32-fc0ca3ec7775" />

### Training with Sliding Windows and CTC Loss

During training, we process each full recitation (e.g., a 30‑second utterance) by sliding a window over the audio with a step size of **320 ms**. For every window position, the model receives **500 ms of past context**, the **current 320 ms chunk**, and **320 ms of look‑ahead** audio. This yields a sequence of $N = \lceil T / \Delta \rceil$ overlapping inputs (approximately 94 windows for 30 seconds). The CTC head produces a phoneme probability sequence for each window, to which we explicitly append a blank token to mark the boundary. All such extended sequences are then concatenated into a single continuous prediction $\mathbf{P}_{\text{conc}}$ that spans the entire utterance. The standard CTC loss is finally computed between $\mathbf{P}_{\text{conc}}$ and the full reference phoneme sequence, allowing the model to be trained end‑to‑end while preserving the streaming‑friendly, window‑based processing of the inference pipeline.

### Notation and Parameters

| Symbol | Value | Description |
|--------|-------|-------------|
| $T_{\text{total}}$ | 30 s | Total duration of a training utterance |
| $\Delta$ | 320 ms | Chunk size / sliding‑window step |
| $C$ | 500 ms | Context window (past audio) |
| $L$ | 320 ms | Look‑ahead window (future audio) |
| $N$ | $\lceil T_{\text{total}} / \Delta \rceil$ | Number of windows (e.g., 94 for 30 s) |
| $\mathbf{x}$ | – | Raw audio waveform of the utterance |
| $\mathbf{y}$ | – | Ground‑truth phoneme sequence (without blanks) |
| $\mathcal{V}$ | – | Vocabulary of phoneme labels, including the blank token $\langle b\rangle$ |
| $f_{\text{ssl}}$ | – | SSL encoder (wav2vec2‑base) |
| $f_{\text{rnn}}$ | – | Memory RNN (maintains hidden state across windows) |
| $f_{\text{ctc}}$ | – | CTC linear projection head |

---

### Forward Pass with Stateful Memory RNN

For each window $n$, we compute:

$$
\mathbf{h}_n^{\text{ssl}} = f_{\text{ssl}}(\mathbf{x}_n)
$$

$$
\mathbf{s}_n = f_{\text{rnn}}(\mathbf{h}_n^{\text{ssl}}, \mathbf{s}_{n-1})
$$

$$
\mathbf{z}_n = f_{\text{ctc}}(\mathbf{s}_n)
$$

After applying softmax over $\mathcal{V}$, we obtain:

$$
\mathbf{p}_n = \text{softmax}(\mathbf{z}_n) \quad \in \mathbb{R}^{T_n \times |\mathcal{V}|}
$$

where $T_n$ is the number of output time steps for that window.

---

### Sequence Concatenation with Blank Insertion

We append a mandatory blank token $\mathbf{b}$ (one‑hot for $\langle b\rangle$) to each window:

$$
\tilde{\mathbf{p}}_n = [\mathbf{p}_n; \mathbf{b}] \quad \in \mathbb{R}^{(T_n + 1) \times |\mathcal{V}|}
$$

Then concatenate across all windows:

$$
\mathbf{P}_{\text{conc}} = \tilde{\mathbf{p}}_1 \oplus \tilde{\mathbf{p}}_2 \oplus \cdots \oplus \tilde{\mathbf{p}}_N \quad \in \mathbb{R}^{T_{\text{total}}^{\text{CTC}} \times |\mathcal{V}|}
$$

where $T_{\text{total}}^{\text{CTC}} = \sum_{n=1}^N (T_n + 1)$.

The log‑probabilities are:

$$
\log \mathbf{P}_{\text{conc}} = \log(\mathbf{P}_{\text{conc}})
$$

---

### CTC Loss

$$
\mathcal{L}_{\text{CTC}} = -\log p(\mathbf{y} \mid \log \mathbf{P}_{\text{conc}})
$$

### **Training & Data Improvements**
1. **Data Augmentation Review**  
   - Evaluate current augmentations [here](https://github.com/obadx/prepare-quran-dataset/blob/a1138eb52c26fc76873e66afd711c36e1cb62ec0/train.py#L279) by testing combinations with human annotators.
   - Remove augmentations that are not perceptible to humans.

2. **Training Data Mix**  
   - Use 60% full speech samples + 40% clipped samples (using forced alignment from the current model).

---

### **References**
- **Current Training Setup:**  
  [Training script](https://github.com/obadx/prepare-quran-dataset/blob/main/train.py) | [Config](https://github.com/obadx/prepare-quran-dataset/blob/main/train_config.yml)
- **Lightweight Conformer Reference:**  
  [Nvidia STT Ar FastConformer (100M)](https://huggingface.co/nvidia/stt_ar_fastconformer_hybrid_large_pcd_v1.0) – [Paper](https://arxiv.org/pdf/2305.05084) | [Code](https://github.com/NVIDIA-NeMo/NeMo/blob/main/nemo/collections/asr/models/hybrid_rnnt_ctc_bpe_models.py#L30)
- **Forced Alignment Guide:**  
  [PyTorch Audio Tutorial](https://docs.pytorch.org/audio/stable/tutorials/forced_alignment_tutorial.html)

---

### **Phase 1: Research & Design** [DONE]
We will use [wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base) 94M parameters
1. Review streaming Conformer variants (e.g., Causal Conformer, Streaming Conformer with memory).
2. Define the exact streaming mechanism (e.g., chunk-wise processing with cross-chunk state retention).
3. Finalize model parameter budget (~100M–200M) and compute requirements.

### **Phase 2: Pretraining**
1. We will pretrain wav2vec2-base on [Muaalem-v3 dataset](https://huggingface.co/datasets/obadx/muaalem-annotated-v3) along side wtih [Tarteel tlog dataset](https://huggingface.co/datasets/tarteel-ai/tlog) totaling 2k hours.
2. We will remove samples <  seconds from tlog
3. for samples lrager than 20 seoconds we will split them but not removing them

### **Phase 3 Fine-tuning Data Preparation**
1. Audit current augmentation pipeline; run human evaluation to filter ineffective augmentations.
2. Apply forced alignment using the current model to generate clipped segments. [SKIP THIS STEP]
3. Prepare final dataset mix (60% full + 40% aligned clips). [SKIP THIS STEP]
4. Add people and noises as data augmentation [here](https://yuan-manx.github.io/ai-audio-datasets/#se)
5. Train on complete silence tracks and fill them with blank token. Datasets used from VAD datasets
6. Add new token `</s>` to represent pause or وقف

### **Phase 3 Fine-tuning**
1. We will fine-tune with our modified memory architecture using wav2vec2-base as ssl.

### **Phase 5: Evaluation & Iteration**
1. Validate both on the Testset [here](https://github.com/obadx/prepare-quran-dataset/blob/a1138eb52c26fc76873e66afd711c36e1cb62ec0/train_config.yml#L29) and the qdat_benchmark [here](https://huggingface.co/datasets/obadx/qdat_bench).
2. Compare against baseline models (Quran Muaalem v3_2) [here](https://huggingface.co/obadx/muaalem-model-v3_2).
3. Conduct human evaluation for quality and naturalness.

**Comment by obadx** [comment]
This is a source of Audio datasets where we can add people and other kinds of background noises [here](https://github.com/Yuan-ManX/ai-audio-datasets)
